operator/pkg/clusterconfiguration/configuration_cluster.go

```diff
@@ -160,16 +160,30 @@ func (c *clusterCfg) finalize(parent *CombinedCfg) error {
 			} else {
 				c.AddFixup(k, fmt.Sprintf(`%s(%s("%s"))`, CELRepr, CELEnvString, envName))
 			}
-		case v.ExternalSecretRef != nil:
-			c.templated[k] = `""`
+		case v.ExternalSecretRef != nil: // nolint:staticcheck // ignore deprecation for now
+			v.ExternalSecretRefSelector = &vectorizedv1alpha1.ExternalSecretKeySelector{
+				Name: *v.ExternalSecretRef, // nolint:staticcheck // ignore deprecation for now
+			}
+			fallthrough
+		case v.ExternalSecretRefSelector != nil:
+			optional := ptr.Deref(v.ExternalSecretRefSelector.Optional, false)
+			switch {
 			// We assume by default that the supplied value is a raw string, which can and should be quoted for the safe
 			// insertion into a bootstrap template.
 			// If that's not the case, and the referred value's octets should be injected into the template verbatim,
 			// then the user can specify that explicitly.
-			if v.UseRawValue {
-				c.AddFixup(k, fmt.Sprintf(`%s("%s")`, CELEnvString, *v.ExternalSecretRef))
-			} else {
-				c.AddFixup(k, fmt.Sprintf(`%s(%s("%s"))`, CELRepr, CELEnvString, *v.ExternalSecretRef))
+			// We wrap the returned vaue in `errorToWarning` in the case where the key is marked as optional.
+			case v.UseRawValue && !optional:
+				c.templated[k] = `""`
+				c.AddFixup(k, fmt.Sprintf(`%s("%s")`, CELExternalSecretRef, v.ExternalSecretRefSelector.Name))
+			case !v.UseRawValue && !optional:
+				c.templated[k] = `""`
+				c.AddFixup(k, fmt.Sprintf(`%s(%s("%s"))`, CELRepr, CELExternalSecretRef, v.ExternalSecretRefSelector.Name))
+			case v.UseRawValue && optional:
+				c.AddFixup(k, fmt.Sprintf(`%s(%s("%s"))`, CELErrorToWarning, CELExternalSecretRef, v.ExternalSecretRefSelector.Name))
```

Comment:
This is the one case where `c.templated[k]` isn't present. After reading this a few times, I _think_ that is intentional. Perhaps this logic could be made more clear by making it more sequential?

```go
expr := fmt.Sprintf(`%s(%q)`, CELExternalSecretRef, v.ExternalSecretRefSelector.Name)

if v.UseRawValue {
    c.templated[k] = `""`
    expr = fmt.Sprintf(`%s(%s)`, CELRepr, expr)
}

if optional {
   expr = fmt.Sprintf(`%s(%s)`, CELErrorToWarning, expr)
}

c.AddFixup(k, expr)
```

Any chance we could create the CEL AST structs here and finalize it into fixups by calling a ToString method?

Reply:
First suggestion looks good.

Second one might need some playing with. I don't want to block Alena so I've left a natural follow-up to this (that is, registering errors and warnings in the cluster's status) for a second PR. I think that change could go in with that too - would that be okay?

Reply:
>  I think that change could go in with that too - would that be okay?

Yeah, np! I was tossing it out just in case it's easier. Sprintf'ing is fine for now.

And on my second re-read I think I flipped the condition of `v.UseRawValue`. It should be `if !v.UseRawValue` IIUC?

Reply:
The addition of the "templated" value is superfluous here. Either the lookup succeeds or it doesn't. If it does, the template value is inserted. If it isn't, then (following Alena's approach) it's left out of the config entirely.